dynamics_controllers/README.md

```diff
@@ -0,0 +1,158 @@
+
+position_controller:
+  type: dynamics_controllers/KdlChainController
+  sub_controller:
+    # the subcontroller is basically the controller you had before!
+    type: effort_controllers/JointGroupPositionController
+    joints:
+      - joint1
+      - joint2
+    joint1/pid: {p: 10.0, i: 0.0, d: 5.0}
+    joint2/pid: {p: 10.0, i: 0.0, d: 5.0}
+
+velocity_controller:
+  type: dynamics_controllers/KdlTreeController
+  sub_controller:
+    # the subcontroller is basically the controller you had before!
+    type: effort_controllers/JointGroupVelocityController
+    joints:
+      - joint1
+      - joint2
+    joint1/pid: {p: 10.0, i: 0.0, d: 5.0}
+    joint2/pid: {p: 10.0, i: 0.0, d: 5.0}
+```
+
+Note that you will likely have to tune again the gains of the sub-controllers to
+achieve good performances.
+
+#### Detailed Controllers Description
+
+Below, parameters relative to each controller are listed.
+Those marked as "searched for" can live in any parent namespace since the
+`NodeHandle::searchParam` method is used to locate them.
+This facilitates "sharing" common parameters such as the gravity vector, which
+should not depend on a specific controller (at least in principle!).
+
+##### KdlChainController
+
+- `sub_controller`: should contain the configuration of a controller of type
+  compatible with `EffortJointInterface`.
+- `robot_description`: should contain the URDF of the robot. Searched for.
+- `gravity`: list with three elements, representing the gravity vector in the
+  base frame of the chain. Searched for. Default: `[0,0,0]`.
+- `chain_base`: base of the chain. Searched for. Default: name of the root link
+  from the robot description.
+- `chain_tip`: tip of the chain. Searched for. Default: if a single "branch"
+  is rooted at `base_link`, `chain_tip` will correspond to the last link of the
+  chain. If at any point multiple children are found, loading will fail.
+- `gravity_compensation_only`: as discussed above. Default: `false`.
+
+##### KdlTreeController
+
+- `sub_controller`: should contain the configuration of a controller of type
+  compatible with `EffortJointInterface`.
+- `robot_description`: should contain the URDF of the robot. Searched for.
+- `gravity`: list with three elements, representing the gravity vector in the
+  base frame of the chain. Searched for. Default: `[0,0,0]`.
+- `tree_root`: base of the tree. Searched for. Default: name of the root link
+  from the robot description.
+- `gravity_compensation_only`: as discussed above. Default: `false`.
+
+
+
+### Limitations
+
+- They all assume a static base. The main implication is that you cannot
+  "serially join" controllers for different parts of the robot.
+- They require proper identification of the Dynamic Parameters.
+- The parameters must be provided in URDF format. This is a limitation since
+  the identification will often return a smaller set of regrouped parameters.
+- The sub-controller must control all joints of the internal `Chain`/`Tree`.
```

Comment:
Couldn't this be detected via parsing the KDL tree?

Reply:
If you mean that it is necessary to make sure that all joints are handled by the sub-controller, then yes, it can be checked from the parsed tree and is addressed in the code [here](https://github.com/francofusco/ros_controllers/blob/1fb22f8473153cdd918f5f34fe5ece857637a512/dynamics_controllers/src/dynamics_controller_base.cpp#L185). My comment in the readme is meant to warn users that it is not possible to run dynamics controllers if the sub-controller evaluates the efforts just for a subset of the joints.